### DISCUSSION TITLE
mypy: missing positional arguments when using default_args

### DISCUSSION THREAD

**Discussion by pavelpi** [apache/airflow#37088] [Q&A]
We are using mypy for static code analysis of Airflow DAGs
When we specify parameters that are injected to each task from DAG's default_args, mypy complains about missing positional arguments [call-arg]

Example:

```
class MyNewOperator(BaseOperator):
    def __init__(self,
                 my_param: int,
                 *args, **kwargs):
		 ...

```

```
default_args = {
    ...
    "my_param": 666
    }

dag = DAG(
    SOME_ID,
    default_args=default_args
)

my_task = MyNewOperator(   # here mypy complains about missing positional argument my_param
    dag=dag,
    task_id="my_new_task"
)
```

Is there some trick how to make mypy happy without using `# type: ignore[call-arg]` ?

**Comment by andresitorusz** [comment]
I don't think that passing other key into `default_args` except the values mentioned in this [docs](https://airflow.apache.org/docs/apache-airflow/1.10.9/tutorial.html#default-arguments) will work.

I guess, you need to create a custom global variable instead of using the `default_args` for your custom parameter.

**Reply by Taragolis** [reply]
It is not limited only by `BaseOperator`'s arguments, and it could be used by any other attributes names, except `task_id`, `dag` and some other. But in general it should work as described into the [Fundamental Concepts: Tasks](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html#tasks)

```python
from __future__ import annotations

from typing import TYPE_CHECKING

from airflow import DAG
from airflow.decorators import task
from airflow.models import BaseOperator
from airflow.utils.task_group import TaskGroup

if TYPE_CHECKING:
    from airflow.utils.context import Context


class CustomOperator(BaseOperator):
    def __init__(self, param_1, param_2, **kwargs):
        super().__init__(**kwargs)
        self.param_1 = param_1
        self.param_2 = param_2

    def execute(self, context: Context):
        print(f"param_1: {self.param_1}")
        print(f"param_2: {self.param_2}")


dag_default_args = {"param_1": "foo", "param_2": "bar"}
tg_default_args = {"param_1": "spam", "param_2": "egg"}

with DAG(
    dag_id="discussion_37088", schedule=None, tags=["37088", "default_args"], default_args=dag_default_args
):
    @task
    def taskflow(param_1: str | None = None) -> None:
        print(f"param_1: {param_1}")

    with TaskGroup(group_id=f"awesome-tg", default_args=tg_default_args) as tg:
        taskflow() >> CustomOperator(task_id="custom")

    taskflow() >> CustomOperator(task_id="custom") >> tg
```

**Comment by Taragolis** [comment]
> Is there some trick how to make mypy happy without using # type: ignore[call-arg] ?

I'm not sure that is possible to do it, because assignments to default args use some hacks into the metaclass [BaseOperatorMeta](https://github.com/apache/airflow/blob/ba053f79edab54366208aad9d88877a95b789eca/airflow/models/baseoperator.py#L364) since Airflow 2.1, or by decorator `apply_defaults` in versions before.

I hope I was wrong and someone could fix signature into the runtime (or type checking) into the `BaseOperatorMeta`

**Comment by potiuk** [comment]
There was an idea from the very beginning that one could write mypy plugin to either check default_args and remove the problems for those or maybe simply apply automatically "call-arg" ignore for DAGs. We would also benefit from that if someone does it. 

So if you would like to fix it @pavlpi - writing and contributing such mypy plugin would be a good idea. (though at the time we discussed it mypy plugins were poorly documented so a bit hacking and reverse engineering was needed - that's why we deferred it.

But maybe situation changed now and it would be a piece of cake now.
